sklearn/preprocessing/tests/test_data.py

```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
 
     assert_raises_regex(ValueError, "axis should be either equal to 0 or 1"
                         ". Got axis=2", quantile_transform, X.T, axis=2)
+
+
+def _generate_random_features_matrix(n_values=3, size=10):
+    rng = np.random.RandomState(0)
+    X = rng.randint(n_values, size=(size, n_values))
+    return X
+
+
+def test_unary_encoder():
+    X = np.arange(5).reshape(-1, 1)
+    enc = UnaryEncoder(5)
+    Xt = enc.fit_transform(X)
+    assert_array_equal(Xt, [[0, 0, 0, 0],   # 0
+                            [1, 0, 0, 0],   # 1
+                            [1, 1, 0, 0],   # 2
+                            [1, 1, 1, 0],   # 3
+                            [1, 1, 1, 1]])  # 4
+    Xt2 = enc.transform(X)
+    assert_array_equal(Xt2, Xt)
+
+
+def test_unary_encoder_stack():
+    # multiple input features stack to same output
+    n_values = np.random.randint(2, 10)
+    size = np.random.randint(1, 10)
+
+    encoder = UnaryEncoder(n_values, sparse=False)
+    X_multi = _generate_random_features_matrix(n_values, size)
+    X_multi_t = encoder.fit_transform(X_multi)
+    assert_equal(X_multi_t.shape, (size, n_values * (n_values - 1)))
+
+    expected = np.hstack([encoder.fit_transform(X_multi[:, i:(i + 1)])
+                          for i in range(X_multi.shape[1])])
+    assert_array_equal(expected, X_multi_t)
+
+
+def test_unary_encoder_dense_sparse():
+    # test dense output in comparison to sparse results.
+    n_values = np.random.randint(1, 10)
+    size = np.random.randint(1, 10)
+
+    sparse_encoder = UnaryEncoder(n_values, sparse=True)
+    dense_encoder = UnaryEncoder(n_values)
+
+    X = _generate_random_features_matrix(n_values, size)
+    X_trans_sparse = sparse_encoder.fit_transform(X)
+    X_trans_dense = dense_encoder.fit_transform(X)
+
+    assert_array_equal(X_trans_sparse.toarray(), X_trans_dense)
+
+
+def test_unary_encoder_handle_greater():
+    X = np.array([[0, 2, 1], [1, 0, 3], [2, 0, 2]])
+    y = np.array([[4, 1, 1]])
+
+    # Test that encoder raises error for greater features.
+    encoder = UnaryEncoder(handle_greater='error')
+    encoder.fit(X)
+    assert_raises(ValueError, encoder.transform, y)
+
+    # Test the ignore option, clips greater features.
+    encoder = UnaryEncoder(handle_greater='clip')
+    encoder.fit(X)
+    assert_array_equal(
+        encoder.transform(y),
+        np.array([[1.,  1.,  1.,  0.,  1.,  0.,  0.]]))
+
+    # Raise error if handle_greater is neither ignore or error.
+    encoder = UnaryEncoder(handle_greater='42')
+    encoder.fit(X)
```

Comment:
All parameter validation should happen in fit. This is where we should expect the error to be raised.